### PR TITLE
[Merged by Bors] - Fix edge-case when finding the finalized descendant

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Description

We were missing an edge case when checking to see if a block is a descendant of the finalized checkpoint. This edge case is described for one of the tests in this PR:

https://github.com/sigp/lighthouse/blob/a119edc739e9dcefe1cb800a2ce9eb4baab55f20/consensus/proto_array/src/proto_array_fork_choice.rs#L1018-L1047

This bug presented itself in the following mainnet log:

```
Jan 26 15:12:42.841 ERRO Unable to validate attestation error: MissingBeaconState(0x7c30cb80ec3d4ec624133abfa70e4c6cfecfca456bfbbbff3393e14e5b20bf25), peer_id: 16Uiu2HAm8RPRciXJYtYc5c3qtCRdrZwkHn2BXN3XP1nSi1gxHYit, type: "unaggregated", slot: Slot(5660161), beacon_block_root: 0x4a45e59da7cb9487f4836c83bdd1b741b4f31c67010c7ae343fa6771b3330489
```

Here the BN is rejecting an attestation because of a "missing beacon state". Whilst it was correct to reject the attestation, it should have rejected it because it attests to a block that conflicts with finality rather than claiming that the database is inconsistent.

The block that this attestation points to (`0x4a45`) is block `C` in the above diagram. It is a non-canonical block in the first slot of an epoch that conflicts with the finalized checkpoint. Due to our lazy pruning of proto array, `0x4a45` was still present in proto-array. Our missed edge-case in [`ForkChoice::is_descendant_of_finalized`](https://github.com/sigp/lighthouse/blob/38514c07f222ff7783834c48cf5c0a6ee7f346d0/consensus/fork_choice/src/fork_choice.rs#L1375-L1379) would have indicated to us that the block is a descendant of the finalized block. Therefore, we would have accepted the attestation thinking that it attests to a descendant of the finalized *checkpoint*.

Since we didn't have the shuffling for this erroneously processed block, we attempted to read its state from the database. This failed because we prune states from the database by keeping track of the tips of the chain and iterating back until we find a finalized block. This would have deleted `C` from the database, hence the `MissingBeaconState` error.

### CODE REVIEW

File: consensus/proto_array/src/proto_array.rs [hunk 1]
```diff
@@ -990,6 +996,70 @@ impl ProtoArray {
...
+            if let Some(parent) = node.parent.and_then(|index| self.nodes.get(index)) {
+                node = parent
+            } else {
+                // If `node` is not the finalized block and its parent does not
+                // exist in fork choice, then the parent must have been pruned.
+                // Proto-array only prunes blocks prior to the finalized block,
+                // so this means the parent conflicts with finality.
+                return false;
+            };
+        }
```
Comment (michaelsproul) [inline consensus/proto_array/src/proto_array.rs:1060]: I think this `loop {}` could be written more similarly to `is_descendant` and pulled out into its own function `is_descendant_from_ancestor_at_slot`:

```rust
/// Check if the ancestor of `descendant_root` at `ancestor_slot` is `ancestor_root`.
///
/// Takes into account ancestors at skipped slots.
fn is_descendant_from_ancestor_at_slot(&self,
    ancestor_root: Hash256,
    ancestor_slot: Slot,
    descendant_root: Hash256
) -> bool {
    self.iter_block_roots(&descendant_root)
        .find(|(_root, slot)| slot <= ancestor_slot)
        .map_or(false, |(root, _)| root == ancestor_root)
}
```

Then we call it as:

```rust
self.is_descendant_from_ancestor_at_slot(finalized_root, finalized_slot, root)
```

(we _should not_ use `node.parent` as that would trigger the exact case being addressed).

Or we could inline it with similar comments? Also happy to keep the longer `loop` version, as it is more explanatory.

File: consensus/proto_array/src/proto_array.rs [hunk 2]
```diff
@@ -990,6 +996,70 @@ impl ProtoArray {
...
+            return false;
+        };
+
+        // The finalized and justified checkpoints represent a list of known
+        // ancestors of `node` that are likely to coincide with the store's
+        // finalized checkpoint.
+        //
+        // Don't continue checking these values for ancestors. If they don't
+        // match for the child then they're unlikely to start matching for its
+        // ancestors.
```
Comment (michaelsproul) [inline consensus/proto_array/src/proto_array.rs]: This comment confused me a bit, is it leftover from a previous implementation that checked these checkpoints on the ancestors?

**Review by michaelsproul — APPROVED** [review]
Great find and nice fix!

Just some quibbling from me about style, but otherwise happy to merge.

File: consensus/proto_array/src/proto_array.rs (hunk 2 above)
Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs]: It's more of a reasoning why these checks are outside the loop, rather than inside it. Happy to re-word if you can show me which part was most confusing :)

File: consensus/proto_array/src/proto_array.rs (hunk 1 above)
Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs:1060]: The only benefit that the current approach has is that it avoids doing a double read of the first `node`. It's certainly not a big deal since it is just hashmap access + array access.

I definitely like the generalized approach and implemented it, but became a little torn when I noticed the double-read 🤔

File: consensus/proto_array/src/proto_array.rs (hunk 2 above)
Comment (michaelsproul) [reply consensus/proto_array/src/proto_array.rs]: Ohhhh I get it now, I took it without the context of the loop below

Comment (michaelsproul) [reply consensus/proto_array/src/proto_array.rs]: Happy to leave as-is

Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs]: I tweaked the comment to avoid this confusion in the future:

```
        // Run this check once, outside of the loop rather than inside the loop.
        // If the conditions don't match for this node then they're unlikely to
        // start matching for its ancestors.
```

File: consensus/proto_array/src/proto_array.rs (hunk 1 above)
Comment (paulhauner) [reply consensus/proto_array/src/proto_array.rs:1060]: After coming back to this I'm tempted to leave the loop as-is. I'm very much on the fence so happy to make the change if you feel strongly ☺️

Comment (michaelsproul) [reply consensus/proto_array/src/proto_array.rs:1060]: Yeah I'm OK to leave as-is. We can refactor in future if we find we want that descendant at slot function for anything else

**Comment by michaelsproul** [comment]
batching

bors r+
